File: client/src/main/java/net/ServerFacade.java
package net;

import java.util.Collection;

import chess.ChessMove;
import model.AuthData;
import model.GameData;
import model.UserData;
import request.CreateGameRequest;
import request.JoinGameRequest;
import response.CreateGameResponse;
import response.ListGamesResponse;
import websocket.commands.ConnectCommand;
import websocket.commands.LeaveCommand;
import websocket.commands.MakeMoveCommand;
import websocket.commands.ResignCommand;

public class ServerFacade {

  private final HttpCommunicator httpCommunicator;
  private final ServerMessageObserver serverMessageObserver;
  private final String domainName;
  private WebsocketCommunicator websocketCommunicator = null;
  private String token;

  public ServerFacade(String domainName, ServerMessageObserver serverMessageObserver) {
    httpCommunicator = new HttpCommunicator("http://" + domainName);
    this.serverMessageObserver = serverMessageObserver;
    this.domainName = domainName;
    token = null;
  }

  public void register(String username, String password, String email) throws ResponseException {
    UserData userData = new UserData(username, password, email);
    AuthData authDataResponse = httpCommunicator.makeRequest("POST", "/user", userData, AuthData.class, token);
    token = authDataResponse.authToken();
  }

  public void login(String username, String password) throws ResponseException {
    UserData userData = new UserData(username, password, null);
    AuthData authDataResponse = httpCommunicator.makeRequest("POST", "/session", userData, AuthData.class, token);
    token = authDataResponse.authToken();
  }

  public void logout() throws ResponseException {
    httpCommunicator.makeRequest("DELETE", "/session", null, null, token);
    token = null;
  }

  public Collection<GameData> listGames() throws ResponseException {
    ListGamesResponse listGamesResponse = httpCommunicator.makeRequest(
      "GET", 
      "/game", 
      null, 
      ListGamesResponse.class, 
      token
    );
    
    return listGamesResponse.games;
  }

  public Integer createGame(String gameName) throws ResponseException {
    CreateGameRequest createGameRequest = new CreateGameRequest(gameName);
    CreateGameResponse response = httpCommunicator.makeRequest(
      "POST", 
      "/game", 
      createGameRequest, 
      CreateGameResponse.class, 
      token
    );
    return response.getGameID();
  }

  public void joinGame(Integer gameID, String teamColor) throws ResponseException {
    JoinGameRequest joinGameRequest = new JoinGameRequest(teamColor, gameID);
    httpCommunicator.makeRequest("PUT", "/game", joinGameRequest, null, token);
    joinObserver(gameID);
  }

  public void joinObserver(Integer gameID) throws ResponseException {
    ConnectCommand connectCommand = new ConnectCommand(token, gameID);
    getWebsocketCommunicator().send(connectCommand);
    
  }

  public void leaveGame(Integer gameID) throws ResponseException {
    LeaveCommand leaveCommand = new LeaveCommand(token, gameID);
    getWebsocketCommunicator().send(leaveCommand);
  }

  public void makeMove(Integer gameID, ChessMove chessMove) throws ResponseException {
    MakeMoveCommand makeMoveCommand = new MakeMoveCommand(token, gameID, chessMove);
    getWebsocketCommunicator().send(makeMoveCommand);
  }

  public void resign(Integer gameID) throws ResponseException {
    ResignCommand resignCommand = new ResignCommand(token, gameID);
    getWebsocketCommunicator().send(resignCommand);
  }

  private WebsocketCommunicator getWebsocketCommunicator() throws ResponseException {
    try {
      if (websocketCommunicator == null) {
        websocketCommunicator = new WebsocketCommunicator("ws://" + domainName, serverMessageObserver);
      }
      return websocketCommunicator;
    } catch (Exception e) {
      throw new ResponseException(e.getMessage());
    }
  }
}

File: react-client/src/net/ServerFacade.js
import { HttpCommunicator } from "./HttpCommunicator";


export class ServerFacade {

  #serverUrl = "http://localhost:8080";
  #httpCommunicator = new HttpCommunicator();
  #authToken = "";

  #gameIdMappings = {};

  /**
   * 
   * @param {string} username 
   * @param {string} password 
   */
  async doLogin(username, password) {
    const jsonBody = {
      "username": username,
      "password": password
    };
    const authData = await this.#httpCommunicator.makeRequest("POST", this.#serverUrl+"/session", jsonBody, null);
    this.#authToken = authData["authToken"];
  }

  /**
   * 
   * @param {string} username 
   * @param {string} password 
   * @param {string} email 
   */
  async doRegister(username, password, email) {
    const jsonBody = {
      "username": username,
      "password": password,
      "email": email
    };
    const authData = await this.#httpCommunicator.makeRequest("POST", this.#serverUrl+"/user", jsonBody, null);
    this.#authToken = authData["authToken"];
  }

  async doLogout() {
    await this.#httpCommunicator.makeRequest("DELETE", this.#serverUrl+"/session", null, this.#authToken);
    this.#authToken = null;
  }

  async doListGames() {
    const gamesListJsonObj = await this.#httpCommunicator.makeRequest("GET", this.#serverUrl+"/game", null, this.#authToken); 
    let counter = 1;
    for (const gameElem of gamesListJsonObj["games"]) {
      this.#gameIdMappings[counter] = gameElem["gameID"];
      gameElem["gameID"] = counter;
      counter++;
    }
    gamesListJsonObj["games"].sort((gameOne, gameTwo) => gameOne["gameName"] < gameTwo["gameName"]);
    return gamesListJsonObj["games"];
  }

  /**
   * 
   * @param {string} gameName 
   */
  async doCreateGame(gameName) {
    const jsonBody = {
      "gameName": gameName
    }
    await this.#httpCommunicator.makeRequest("POST", this.#serverUrl+"/game", jsonBody, this.#authToken);
  }

  /**
   * 
   * @param {number} gameID 
   * @param {string} teamColor 
   */
  async doJoinGame(gameID, teamColor) {
    const realGameID = this.#gameIdMappings[gameID];
    const jsonBody = {
      "playerColor": teamColor,
      "gameID": realGameID
    }
    await this.#httpCommunicator.makeRequest("PUT", this.#serverUrl+"/game", jsonBody, this.#authToken);
  }

  /**
   * 
   * @param {number} gameID 
   */
  async doObserveGame(gameID) {

  }
}